README.md
===================
biodec.recipe.riak
===================

zc.buildout recipe for downloading and installing Riak.

[![Build Status](https://secure.travis-ci.org/biodec/biodec.recipe.riak.png)](http://travis-ci.org/biodec/biodec.recipe.riak)

see src/biodec/recipe/biodec/README.txt

Credits
=======

Code from http://pypi.python.org/pypi/rod.recipe.rabbitmq package has been used
as a starting point.


CHANGES.txt
Changes
=======

1.0.0a2 (unreleased)
--------------------

- riak 1.2.1, no changes on this recipe [mauro]


1.0.0a1 (2012-10-29)
--------------------

- first public release.


src/biodec/recipe/riak/tests/test_doc.py
"""Doctest runner for biodec.recipe.riak."""

__docformat__ = 'restructuredtext'

import unittest
import zc.buildout.tests
import zc.buildout.testing

from zope.testing import doctest, renormalizing

optionflags = (doctest.ELLIPSIS |
               doctest.NORMALIZE_WHITESPACE |
               doctest.REPORT_ONLY_FIRST_FAILURE)


def setUp(test):
    zc.buildout.testing.buildoutSetUp(test)
    zc.buildout.testing.install_develop('zc.recipe.egg', test)
    zc.buildout.testing.install_develop('biodec.recipe.riak', test)
    zc.buildout.testing.install_develop('simplejson', test)


def test_suite():
    suite = unittest.TestSuite((
            doctest.DocFileSuite(
                '../README.txt',
                setUp=setUp,
                tearDown=zc.buildout.testing.buildoutTearDown,
                optionflags=optionflags,
                checker=renormalizing.RENormalizing([
                        zc.buildout.testing.normalize_path,
                        ]),
                ),
            ))
    return suite


if __name__ == '__main__':
    unittest.main(defaultTest='test_suite')


src/biodec/recipe/riak/build.py
# -*- coding: utf-8 -*-
# Copyright (C)2012 'Biodec'

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
"""Recipe for setting up Riak."""

import logging
import os
import pkg_resources
import shutil
import subprocess
import sys
import tempfile
import urllib
import zc.recipe.egg

logger = logging.getLogger(__name__)


class BuildRecipe(zc.recipe.egg.Eggs):
    """Buildout recipe for installing Riak."""

    def __init__(self, buildout, name, opts):
        """Standard constructor for zc.buildout recipes."""

        super(BuildRecipe, self).__init__(buildout, name, opts)
        self.options['location'] = os.path.join(
            self.buildout['buildout']['parts-directory'],
            self.name)

    def install_riak(self):
        """Downloads and installs Riak."""

        arch_filename = self.options['url'].split(os.sep)[-1]
        dst = self.options['location']
        # Re-use the buildout download cache if defined
        downloads_dir = self.buildout['buildout'].get('download-cache')
        if downloads_dir is None:
            downloads_dir = os.path.join(self.buildout['buildout']['directory'],
                                        'downloads')
            if not os.path.isdir(downloads_dir):
                os.mkdir(downloads_dir)
            self.buildout['buildout'].setdefault('download-cache', downloads_dir)
        logger.info("downloading Riak distribution... %r %r" % (downloads_dir, arch_filename))
        src = os.path.join(downloads_dir, arch_filename)
        if not os.path.isfile(src):
            logger.info("downloading Riak distribution...")
            urllib.urlretrieve(self.options['url'], src)
        else:
            logger.info("Riak distribution already downloaded.")

        extract_dir = tempfile.mkdtemp("buildout-" + self.name)
        remove_after_install = [extract_dir]
        is_ext = arch_filename.endswith
        is_archive = True
        if is_ext('.tar.gz') or is_ext('.tgz'):
            call = ['tar', 'xzf', src, '-C', extract_dir]
        elif is_ext('.zip'):
            call = ['unzip', src, '-d', extract_dir]
        else:
            is_archive = False

        if is_archive:
            retcode = subprocess.call(call)
            if retcode != 0:
                raise Exception("extraction of file %r failed (tempdir: %r)" %
                                (arch_filename, extract_dir))
        else:
            shutil.copy(arch_filename, extract_dir)

        if is_archive:
            top_level_contents = os.listdir(extract_dir)
            if len(top_level_contents) != 1:
                raise ValueError("can't strip top level directory because "
                                 "there is more than one element in the "
                                 "archive.")
            base = os.path.join(extract_dir, top_level_contents[0])
        else:
            base = extract_dir

        if not os.path.isdir(dst):
            os.mkdir(dst)

            for filename in os.listdir(base):
                shutil.move(os.path.join(base, filename),
                            os.path.join(dst, filename))
        else:
            logger.info("Riak already installed.")

        r = pkg_resources.working_set.find(
            pkg_resources.Requirement.parse('simplejson'))

        os.environ['PYTHONPATH'] = r.location
        erlang_path = self.options.get('erlang-path')
        if erlang_path:
            new_path = [erlang_path] + os.environ['PATH'].split(':')
            os.environ['PATH'] = ':'.join(new_path)

        old_cwd = os.getcwd()
        os.chdir(dst)
        retcode = subprocess.call(['make', 'PYTHON=%s' % sys.executable])
        if retcode != 0:
            raise Exception("building Riak failed")
        os.chdir(old_cwd)
    
        for path in remove_after_install:
            shutil.rmtree(path)

        return [dst,]

    def install(self):
        """Creates the part."""

        return self.install_riak()

    def update(self):
        pass


src/biodec/recipe/riak/__init__.py
"""Recipe for setting up Riak."""

import logging
import os
import pkg_resources
import shutil
import subprocess
import sys
import tempfile
import urllib
import zc.recipe.egg

logger = logging.getLogger(__name__)


class Recipe(zc.recipe.egg.Eggs):
    """Buildout recipe for installing Riak."""

    def __init__(self, buildout, name, opts):
        """Standard constructor for zc.buildout recipes."""

        super(Recipe, self).__init__(buildout, name, opts)

    def gen_scripts(self):
        """Generates Riak bin scripts."""
        # TODO

        buildout = self.buildout['buildout']['directory']
        bindir = self.buildout['buildout']['bin-directory']
        part = self.options['location']

        erlang_path = self.options.get('erlang-path')
        if erlang_path:
            erlang_path = 'PATH=%s:$PATH' % erlang_path
        else:
            erlang_path = ''

        return []

    def install_riak(self):
        """Downloads and installs Riak."""

        arch_filename = self.options['url'].split(os.sep)[-1]
        dst = self.options.setdefault('location',
            os.path.join(self.buildout['buildout']['parts-directory'],
            self.name))
        # Re-use the buildout download cache if defined
        downloads_dir = self.buildout['buildout'].get('download-cache')
        if downloads_dir is None:
            downloads_dir = os.path.join(self.buildout['buildout']['directory'],
                                        'downloads')
            if not os.path.isdir(downloads_dir):
                os.mkdir(downloads_dir)
            self.buildout['buildout'].setdefault('download-cache', downloads_dir)
        logger.info("downloading Riak distribution... %r %r" % (downloads_dir, arch_filename))
        src = os.path.join(downloads_dir, arch_filename)
        if not os.path.isfile(src):
            logger.info("downloading Riak distribution...")
            urllib.urlretrieve(self.options['url'], src)
        else:
            logger.info("Riak distribution already downloaded.")

        extract_dir = tempfile.mkdtemp("buildout-" + self.name)
        remove_after_install = [extract_dir]
        is_ext = arch_filename.endswith
        is_archive = True
        if is_ext('.tar.gz') or is_ext('.tgz'):
            call = ['tar', 'xzf', src, '-C', extract_dir]
        elif is_ext('.zip'):
            call = ['unzip', src, '-d', extract_dir]
        else:
            is_archive = False

        if is_archive:
            retcode = subprocess.call(call)
            if retcode != 0:
                raise Exception("extraction of file %r failed (tempdir: %r)" %
                                (arch_filename, extract_dir))
        else:
            shutil.copy(arch_filename, extract_dir)

        if is_archive:
            top_level_contents = os.listdir(extract_dir)
            if len(top_level_contents) != 1:
                raise ValueError("can't strip top level directory because "
                                 "there is more than one element in the "
                                 "archive.")
            base = os.path.join(extract_dir, top_level_contents[0])
        else:
            base = extract_dir

        if not os.path.isdir(dst):
            os.mkdir(dst)

            for filename in os.listdir(base):
                shutil.move(os.path.join(base, filename),
                            os.path.join(dst, filename))
        else:
            logger.info("Riak already installed.")

        r = pkg_resources.working_set.find(
            pkg_resources.Requirement.parse('simplejson'))

        os.environ['PYTHONPATH'] = r.location
        erlang_path = self.options.get('erlang-path')
        if erlang_path:
            new_path = [erlang_path] + os.environ['PATH'].split(':')
            os.environ['PATH'] = ':'.join(new_path)

        old_cwd = os.getcwd()
        os.chdir(dst)
        retcode = subprocess.call(['make', 'PYTHON=%s' % sys.executable])
        if retcode != 0:
            raise Exception("building Riak failed")
        os.chdir(old_cwd)
        
        paths = self.gen_scripts()

        for path in remove_after_install:
            shutil.rmtree(path)

        return paths + [dst]

    def install(self):
        """Creates the part."""

        return self.install_riak()

    def update(self):
        pass


src/biodec/recipe/riak/instance.py
# -*- coding: utf-8 -*-
# Copyright (C)2012 'Biodec'

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
"""Config Recipe riak"""
import os
import subprocess
import logging

RECIPE_BUILD_NAME = 'biodec.recipe.riak:build'


def get_options_from_build(buildout, options):
    part = options.get('riakbuildpart', None)
    if part:
        return buildout[part]

    for part in buildout.keys():
        if buildout[part].has_key('recipe') and \
            buildout[part]['recipe'] == RECIPE_BUILD_NAME:
            return buildout[part]
    return {}

class InstanceRecipe(object):
    """This recipe is used by zc.buildout"""

    def __init__(self, buildout, name, options):
        self.buildout = buildout
        self.name = name
        location = options.get(
            'location', buildout['buildout']['parts-directory'])
        print location
        options['location'] = os.path.join(location, name)
        options['prefix'] = options['location']

        self.options = options
        self.buildoptions = get_options_from_build(buildout, options)
        self.logger = logging.getLogger(__name__)

    def gen_scripts(self, target_dir):
        """Generates Riak bin scripts."""
        bindir = self.buildout['buildout']['bin-directory']

        erlang_path = self.options.get('erlang-path')
        if erlang_path:
            erlang_path = 'PATH=%s:$PATH' % erlang_path
        else:
            erlang_path = ''
        scripts = []
        for scriptname in  ('riak', 'riak-admin', 'search-cmd'):
            script = os.path.join(bindir, "%s.%s" % (self.name, scriptname))
            f = open(script, 'wb')
            f.write('#!/usr/bin/env bash\n%s\ncd %s\nexec bin/%s $@\n' %
                (erlang_path, target_dir, scriptname))
            print erlang_path, target_dir, scriptname
            f.close()
            os.chmod(script, 0755)
            scripts.append(script)
        return scripts

    def install(self):
        """ install riak instance """
        dst = self.options.setdefault('location',
            os.path.join(self.buildout['buildout']['parts-directory'],
            self.name))
        print 'dst', dst
        var = os.path.join(
            self.buildout['buildout']['directory'],
            'var', self.name)
        print 'var', var
        if not os.path.isdir(dst):
            os.mkdir(dst)
        target_dir = os.path.join(dst, 'rel')
        overlay_vars = os.path.join(dst, 'vars.config')
        open(overlay_vars, 'w').write(CONFIG_TEMPLATE % dict(
            root=target_dir,
            var=var,
            web_ip=self.options.get('web_ip', '127.0.0.1'),
            web_port=self.options.get('web_port', 8098)
        ))
        old_cwd = os.getcwd()
        os.chdir(self.buildoptions['location'])
        my_env = os.environ.copy()
        if self.buildoptions.get('erlang-path'):
            my_env["PATH"] = "%s:%s" % (self.buildoptions.get('erlang-path'), my_env.get("PATH"))
        retcode = subprocess.Popen(
            ['./rebar', 'generate', 'target_dir=%s' % target_dir,'overlay_vars=%s' % overlay_vars], 
            env=my_env).wait()
        if retcode != 0:
            raise Exception("Creating Riak instance %s" % self.name)
        os.chdir(old_cwd)
        
        scripts = self.gen_scripts(target_dir)
        return [dst,] + scripts

    def update(self):
        """ update riak instance """
        self.logger.warning('not implemented')


CONFIG_TEMPLATE = '''
%%%% -*- mode: erlang;erlang-indent-level: 4;indent-tabs-mode: nil -*-
%%%% ex: ft=erlang ts=4 sw=4 et

%%%% Platform-specific installation paths
{platform_bin_dir,  "%(root)s/bin"}.
{platform_data_dir, "%(var)s/data"}.
{platform_etc_dir,  "%(root)s/etc"}.
{platform_lib_dir,  "%(root)s/lib"}.
{platform_log_dir,  "%(var)s/log"}.

%%%%
%%%% etc/app.config
%%%%
{web_ip,            "%(web_ip)s"}.
{web_port,          %(web_port)s}.
{handoff_port,      8099}.
{pb_ip,             "127.0.0.1"}.
{pb_port,           8087}.
{ring_state_dir,    "{{platform_data_dir}}/ring"}.
{bitcask_data_root, "{{platform_data_dir}}/bitcask"}.
{leveldb_data_root, "{{platform_data_dir}}/leveldb"}.
{sasl_error_log,    "{{platform_log_dir}}/sasl-error.log"}.
{sasl_log_dir,      "{{platform_log_dir}}/sasl"}.
{mapred_queue_dir,  "{{platform_data_dir}}/mr_queue"}.

%%%% riak_search
{merge_index_data_root,  "{{platform_data_dir}}/merge_index"}.

%%%% secondary indices
{merge_index_data_root_2i,  "{{platform_data_dir}}/merge_index_2i"}.

%%%% Javascript VMs
{map_js_vms,   8}.
{reduce_js_vms, 6}.
{hook_js_vms, 2}.

%%%%
%%%% etc/vm.args
%%%%
{node,         "riak@127.0.0.1"}.
{crash_dump,   "{{platform_log_dir}}/erl_crash.dump"}.

%%%%
%%%% bin/riak
%%%%
{runner_script_dir,  "$(cd ${0%%/*} && pwd)"}.
{runner_base_dir,    "${RUNNER_SCRIPT_DIR%%/*}"}.
{runner_etc_dir,     "$RUNNER_BASE_DIR/etc"}.
{runner_log_dir,     "$RUNNER_BASE_DIR/log"}.
{pipe_dir,           "$RUNNER_BASE_DIR/tmp/"}.
{runner_user,        ""}.        
'''
